bigbio/hub/hub_repos/chebi/chebi.py

```diff
@@ -0,0 +1,247 @@
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                gen_kwargs={
+                    "file_paths": file_paths,
+                    "split": "train",
+                },
+            ),
+        ]
+
+    def _generate_examples(self, file_paths, split: str) -> Tuple[int, Dict]:
+        """Yields examples as (key, example) tuples."""
+
+        if self.config.schema == "source":
+            for filepath in file_paths:
+                key, example = self._read_example_from_file(filepath)
+                yield key, example
+
+        elif self.config.schema == "bigbio_kb":
+            for filepath in file_paths:
+                key, example = self._read_example_from_file_in_kb_schema(filepath)
+                yield key, example
+
+    def _parse_paragraph(self, para, start=0):
+        para_text = []
+        entities = []
+        for e in para.iter():
+            para_text.append(e.text)
+            if e.tag == "ne":
+                entity = {
+                    "phrase": e.text,
+                    "start": start,
+                    "end": start + len(e.text),
+                    "attrs": dict(e.attrib),
+                }
+                entities.append(entity)
+            start += len(e.text)
+            if e.tail:
+                para_text.append(e.tail)
+                start += len(e.tail)
+        return "".join(para_text), entities
+
+    def _read_example_from_file(self, filepath: Path) -> Tuple[str, Dict]:
+        with open(filepath, encoding="utf-8") as fp:
+            xml = etree.fromstring(fp.read().encode("utf-8"))
+        key = filepath.parent.name
+        document_text = []
+        entities = []
+        start = 0
+
+        for para in xml.iter("P", "p"):
+            para_text, para_entities = self._parse_paragraph(para, start=start)
+            document_text.append(para_text)
+            start += len(para_text)
+            entities.extend(para_entities)
+
+        document_text = "".join(document_text)
+        example = {"doc_id": key, "text": document_text, "entities": entities}
+
+        return key, example
+
+    def _parse_example_to_kb_schema(self, example) -> Dict[str, Any]:
+        text = example["text"]
+        doc_id = example["doc_id"]
+        passages = [
+            {
+                "id": f"{doc_id}-P0",
+                "type": "abstract",
+                "text": [text],
+                "offsets": [[0, len(text)]],
+            }
+        ]
+        entities = []
+        for i, e in enumerate(example["entities"]):
+            entity = {
+                "id": f"{doc_id}-E{i}",
+                "text": [e["phrase"]],
+                "offsets": [[e["start"], e["end"]]],
+                "type": e["attrs"]["type"],
+                "normalized": [
+                    {"db_name": "chebi", "db_id": chebi_id.strip()} for chebi_id in e["attrs"]["chebi-id"].split(",")
```

Comment:
The IDs extracted this way look inconsistent, e.g.:

```
 {'id': 'WO2007000651-E8',
   'type': 'CHEMICAL',
   'text': ['Zinc oxide'],
   'offsets': [[613, 623]],
   'normalized': [{'db_name': 'chebi', 'db_id': 'CHEBI:36560'}]},
  {'id': 'WO2007000651-E9',
   'type': 'ONT',
   'text': ['astringent'],
   'offsets': [[690, 700]],
   'normalized': [{'db_name': 'chebi', 'db_id': 'WO2007000651:157583'}]},
  ```
  
  Maybe the db_id should just contain the last numerical bit?